File: packages/flutter-im-ui/lib/src/models/connection_notice.dart
import 'package:flutter/foundation.dart';

import '../components/flare_status_banner.dart' show FlareStatusTone;
import '../tokens/flare_strings.dart';

/// Where the IM connection stands, as an app shell reports it.
enum FlareConnectionPhase {
  connected,
  connecting,
  reconnecting,
  offline,
  disconnected,
  kicked,
  expired,
}

/// The way back a [FlareConnectionNotice] offers.
enum FlareConnectionRecovery { reconnect, signIn }

/// What an app shell tells the user about its IM connection: the words, the
/// tone and the way back, the same on every app. Render it with
/// `FlareStatusBanner` (text, tone, pulse, and the recovery as its action).
@immutable
class FlareConnectionNotice {
  const FlareConnectionNotice({
    required this.phase,
    required this.text,
    required this.tone,
    this.pulse = false,
    this.recovery,
    this.recoveryText,
  });

  final FlareConnectionPhase phase;
  final String text;

  /// [FlareStatusTone.warning] while the connection can come back on its own
  /// or with a reconnect, [FlareStatusTone.danger] when only signing in again
  /// helps.
  final FlareStatusTone tone;

  /// A connection attempt is under way (the banner's dot pulses).
  final bool pulse;

  /// The action the banner offers, with [recoveryText] as its label; null
  /// offers none.
  final FlareConnectionRecovery? recovery;
  final String? recoveryText;
}

/// The notice for [phase], in [strings]' words; null when connected.
///
/// - connecting and reconnecting say so (the core retries on its own);
/// - offline waits for the network;
/// - disconnected names [reason] when there is one, and offers 重新连接 only
///   when the host [canReconnect];
/// - kicked and expired are final (the core does not reconnect from them), so
///   they always offer 重新登录; a [reason] replaces the kicked text.
FlareConnectionNotice? flareConnectionNotice(
  FlareConnectionPhase phase,
  FlareStrings strings, {
  String? reason,
  bool canReconnect = false,
}) {
  final why = reason?.trim() ?? '';
  return switch (phase) {
    FlareConnectionPhase.connected => null,
    FlareConnectionPhase.connecting => FlareConnectionNotice(
      phase: phase,
      text: strings.connectionConnecting,
      tone: FlareStatusTone.warning,
      pulse: true,
    ),
    FlareConnectionPhase.reconnecting => FlareConnectionNotice(
      phase: phase,
      text: strings.connectionReconnecting,
      tone: FlareStatusTone.warning,
      pulse: true,
    ),
    FlareConnectionPhase.offline => FlareConnectionNotice(
      phase: phase,
      text: strings.connectionOffline,
      tone: FlareStatusTone.warning,
    ),
    FlareConnectionPhase.disconnected => FlareConnectionNotice(
      phase: phase,
      text: why.isEmpty
          ? strings.connectionDisconnected
          : strings.connectionDisconnectedReason(why),
      tone: FlareStatusTone.warning,
      recovery: canReconnect ? FlareConnectionRecovery.reconnect : null,
      recoveryText: canReconnect ? strings.connectionReconnect : null,
    ),
    FlareConnectionPhase.kicked => FlareConnectionNotice(
      phase: phase,
      text: why.isEmpty ? strings.connectionKicked : why,
      tone: FlareStatusTone.danger,
      recovery: FlareConnectionRecovery.signIn,
      recoveryText: strings.connectionSignIn,
    ),
    FlareConnectionPhase.expired => FlareConnectionNotice(
      phase: phase,
      text: strings.connectionExpired,
      tone: FlareStatusTone.danger,
      recovery: FlareConnectionRecovery.signIn,
      recoveryText: strings.connectionSignIn,
    ),
  };
}

File: packages/android-im-ui/src/main/kotlin/com/flare/im/ui/ConnectionNotice.kt
package com.flare.im.ui

import androidx.compose.runtime.Immutable

/** The IM connection states an app shell tells the user about (the Vue `FlareConnectionPhase`). */
enum class FlareConnectionPhase { Connected, Connecting, Reconnecting, Offline, Disconnected, Kicked, Expired }

/** The way back a [FlareConnectionNotice] offers: try the connection again, or sign in again. */
enum class FlareConnectionRecovery { Reconnect, SignIn }

/**
 * What an app shell shows for its connection, drawn with [StatusBanner]: [text] in [tone], a pulsing dot
 * while a connection attempt is under way ([pulse]), and the recovery action labelled [recoveryText].
 */
@Immutable
data class FlareConnectionNotice(
    val phase: FlareConnectionPhase,
    val text: String,
    val tone: FlareStatusTone,
    val pulse: Boolean,
    val recovery: FlareConnectionRecovery? = null,
    val recoveryText: String? = null,
)

/**
 * The notice for [phase] (Vue `connectionNotice`); the kit owns the wording, the tone and the way back, so
 * every app says the same thing for the same state. Connected shows nothing (null). Connecting and
 * reconnecting pulse, since the core retries on its own; offline waits for the network; disconnected names
 * [reason] when there is one and offers to reconnect only when the host [canReconnect]. Kicked and expired
 * are final (the core does not reconnect from them), so both are danger and always offer to sign in again;
 * a kicked [reason] replaces the default text.
 */
fun flareConnectionNotice(
    phase: FlareConnectionPhase,
    strings: FlareStrings,
    reason: String? = null,
    canReconnect: Boolean = false,
): FlareConnectionNotice? {
    val why = reason?.trim().orEmpty()
    return when (phase) {
        FlareConnectionPhase.Connected -> null
        FlareConnectionPhase.Connecting -> FlareConnectionNotice(phase, strings.connectionConnecting, FlareStatusTone.Warning, pulse = true)
        FlareConnectionPhase.Reconnecting -> FlareConnectionNotice(phase, strings.connectionReconnecting, FlareStatusTone.Warning, pulse = true)
        FlareConnectionPhase.Offline -> FlareConnectionNotice(phase, strings.connectionOffline, FlareStatusTone.Warning, pulse = false)
        FlareConnectionPhase.Disconnected -> FlareConnectionNotice(
            phase = phase,
            text = if (why.isNotEmpty()) strings.connectionDisconnectedReason(why) else strings.connectionDisconnected,
            tone = FlareStatusTone.Warning,
            pulse = false,
            recovery = FlareConnectionRecovery.Reconnect.takeIf { canReconnect },
            recoveryText = strings.connectionReconnect.takeIf { canReconnect },
        )
        FlareConnectionPhase.Kicked -> FlareConnectionNotice(
            phase, why.ifEmpty { strings.connectionKicked }, FlareStatusTone.Danger, pulse = false,
            recovery = FlareConnectionRecovery.SignIn, recoveryText = strings.connectionSignIn,
        )
        FlareConnectionPhase.Expired -> FlareConnectionNotice(
            phase, strings.connectionExpired, FlareStatusTone.Danger, pulse = false,
            recovery = FlareConnectionRecovery.SignIn, recoveryText = strings.connectionSignIn,
        )
    }
}
